### PR TITLE
Add `AbstractArray.is_compatible_with` and misc. changes

### PR DESCRIPTION
**Pull request by mudit2812**
* Add weak typing to `AbstractArray`. A boolean attribute is set based on whether the input was a numpy-like dtype or a Python builtin dtype.
  * For weakly typed abstract arrays , concrete inputs that can be _safely promoted_ to the dtype of the abstract array are considered to be compatible. E.g., `int32`, `int64`, `bool_`, `float32`, etc. are compatible with `float64`, but `complex128` is not, as that would be a demotion.
* Fix logic for dynamically shaped abstract types. In the ADR, a single `...` input is meant to denote arbitrary _rank_. `-1` is meant to denote arbitrary size for a given axis. The implementation for `AbstractArray` and `AbstractWire` (and their respective factories) is updated accordingly.
* Add `__repr__` to `AbstractArray`, `AbstractWires`
* Add `AbstractArray.is_compatible_with`. This is a slightly less restrictive version of `isinstance(..., AbstractArray(...))`.
  * This method returns `True` iff the following conditions are met:
    * If the abstract array is weakly typed, the dtype of the input must be _safely promotable_ to the dtype of the abstract array. Else, the dtype must match exactly.
    * If the abstract array can have arbitrary rank (shape=...), inputs with any shape are allowed.
    * If the abstract array has an arbitrary size for any axes (-1), the shape of the input must be equal in all axes other than the arbitrarily-shaped axes of the abstract array.

[sc-123061][sc-122945]

### CODE REVIEW

File: tests/test_typing.py [hunk 1]
```diff
@@ -281,26 +314,92 @@ def test_scalar_shortcuts(self, shortcut, dtype):
...
+        assert isinstance(np.ones((4, 2)), aa)
+        assert isinstance(np.array(0.5), aa)
+        assert not isinstance(np.ones((4, 2), dtype=np.int64), aa)
+
+    def test_instance_check_rejects_non_array(self):
+        """Test that non-array objects without a shape fail ``isinstance`` checks."""
+        aa = AbstractArray((2,), np.float64)
+        assert not isinstance(0.5, aa)
+
+    def test_weak_type_for_number_dtypes(self):
```
Comment (andrijapau) [inline tests/test_typing.py:367]: Wouldn't it be better to test the consequences of this private flag through the public `is_compatible_with`? You're doing it below for `float`, maybe we want another test for the weak type `int` and then a strong type test for float32 and bool_?

File: tests/test_typing.py [hunk 2]
```diff
@@ -281,26 +314,92 @@ def test_scalar_shortcuts(self, shortcut, dtype):
...
+        """Test that number dtypes are marked as weak types."""
+        assert AbstractArray((), float)._weak_type is True
+        assert AbstractArray((2, 3), int)._weak_type is True
+        assert AbstractArray((2, 3), np.float32)._weak_type is False
+        assert AbstractArray((2, 3), np.bool_)._weak_type is False
+
+    def test_is_compatible_with_weak_scalar(self):
+        """Test ``is_compatible_with`` for scalar values."""
+        aa = AbstractArray((), float)
+        assert aa._weak_type is True
```
Comment (andrijapau) [inline tests/test_typing.py]: I feel like the asserts below show that it's a weak type
```suggestion
```

File: tests/test_typing.py [hunk 3]
```diff
@@ -281,26 +314,92 @@ def test_scalar_shortcuts(self, shortcut, dtype):
...
+        assert aa.is_compatible_with(0.5)
+        assert aa.is_compatible_with(np.array(0.5))
+        assert aa.is_compatible_with(1)
+        assert not aa.is_compatible_with(np.array([0.5, 0.6]))
+        assert not aa.is_compatible_with(1 + 0j)
+
+    def test_is_compatible_with_weak_array(self):
+        """Test ``is_compatible_with`` for array values."""
+        aa = AbstractArray((2, 3), float)
+        assert aa._weak_type is True
```
Comment (andrijapau) [inline tests/test_typing.py]:
```suggestion
```

File: pennylane/typing.py [hunk 4]
```diff
@@ -137,79 +137,154 @@ class AbstractArray:
...
+        if self.shape is Ellipsis:
+            raise TypeError(f"ndim is undefined for {self} with incomplete shape.")
         return len(self.shape)
 
+    def _resolve_dtype(self, dtype):
+        """Convert an arbitrary dtype into a numpy dtype."""
+        if dtype.__class__.__module__.split(".")[0] == "torch":
+            import torch  # pylint: disable=import-outside-toplevel
+
+            dummy = torch.tensor((), dtype=dtype)
```
Comment (andrijapau) [inline pennylane/typing.py]: Probably better to use empty rather than tensor for the purpose of a dummy tensor,
```suggestion
            dummy = torch.empty((), dtype=dtype)
```

> The advantage of torch.empty is that it’s faster, because it will just assign the memory to the tensor without filling it with specific values. [[source](https://discuss.pytorch.org/t/difference-between-torch-tensor-and-torch-empty/32831/6)]

File: pennylane/typing.py [hunk 5]
```diff
@@ -137,79 +137,154 @@ class AbstractArray:
...
+            dtype = dummy.numpy().dtype
+
+        return np.dtype(dtype)
+
+    def __repr__(self):
+        shape_repr = "?" if self.shape is Ellipsis else self.shape
```
Comment (andrijapau) [inline pennylane/typing.py]: I'd be up for keeping `...` in the `repr`. Less logic needed and is representative of the true data.

**Review by andrijapau — COMMENTED** [review]
Looks great! Just some comments after an initial pass 😃

File: pennylane/typing.py (hunk 4 above)
Comment (mudit2812) [reply pennylane/typing.py]: Sure, but I don't think it matters for a tensor with size=0. There is no memory to fill

File: pennylane/typing.py (hunk 5 above)
Comment (mudit2812) [reply pennylane/typing.py]: I wasn't sure what to use here tbh. If `shape is Ellipsis`, then it prints as `Ellipsis`, not `...` 😅 . It's easy enough to fix by doing this though:
```suggestion
        shape_repr = "..." if self.shape is Ellipsis else self.shape
```

Comment (andrijapau) [reply pennylane/typing.py]: Huh. Interesting - I'd be good with that!

File: tests/test_typing.py [hunk 6]
```diff
@@ -171,27 +173,62 @@ def test_hash_and_equality(self):
...
+        assert repr(a0) == "AbstractArray((1, 2), int64, weak_type=True)"
+
+        a1 = AbstractArray((1, 2), np.int32)
+        assert repr(a1) == "AbstractArray((1, 2), int32)"
+
+        a2 = AbstractArray((-1, 2), np.int32)
+        assert repr(a2) == "AbstractArray((-1, 2), int32)"
+
+        a3 = AbstractArray(..., np.int32)
+        assert repr(a3) == "AbstractArray(?, int32)"
```
Comment (mudit2812) [inline tests/test_typing.py]:
```suggestion
        assert repr(a3) == "AbstractArray(..., int32)"
```

File: tests/test_typing.py (hunk 1 above)
Comment (mudit2812) [reply tests/test_typing.py:367]: I'm wondering if it makes sense to make `weak_type` public. I don't think anything outside `AbstractArray` would care about it, but maybe that is my naivety. I don't want to allow _setting_ it in the constructor, but I don't see why it can't be public

File: pennylane/typing.py
Comment (andrijapau) [inline pennylane/typing.py:1]: GitHub is rendering it as the entire file has changed. Is this an EOL problem?  🤔

Comment (mudit2812) [reply pennylane/typing.py:1]: Fixed. Idk what happened

File: tests/test_typing.py (hunk 1 above)
Comment (andrijapau) [reply tests/test_typing.py:367]: Yeah, I think keep it private for now. 🤷🏼

File: pennylane/typing.py [hunk 7]
```diff
@@ -137,79 +137,161 @@ class AbstractArray:
...
-        object.__setattr__(self, "dtype", np.dtype(self.dtype))
+            for s in shape:
+                if not isinstance(s, int) or s < -1:
+                    raise ValueError(
+                        f"Shapes can only be initialized with integer values, but got {shape}. "
+                        "For axes with unknown sizes, use -1. For an unknown number of axes, "
+                        "use shape=Ellipsis."
+                    )
+                if s == -1:
+                    shape_fixed = False
```
Comment (astralcai) [inline pennylane/typing.py:191]:
```suggestion
            if any(not isinstance(s, int) or s < -1 for s in shape):
                raise ValueError(
                    f"Shapes can only be initialized with integer values, but got {shape}. "
                    "For axes with unknown sizes, use -1. For an unknown number of axes, "
                    "use shape=Ellipsis."
                )

            shape_fixed = all(s != -1 for s in shape)
```
This is purely a personal stylistic preference, feel free to use it or ignore it.

File: pennylane/typing.py [hunk 8]
```diff
@@ -137,79 +137,161 @@ class AbstractArray:
...
+        dtype = getattr(instance, "dtype", None)
+        if dtype is None or self._resolve_dtype(dtype) != self.dtype:
             return False
+
+        if self.shape is Ellipsis:
+            return True
+
         shape = getattr(instance, "shape", None)
-        if shape is None or len(shape) != len(self.shape):
+        return shape is not None and self._shape_matches(shape)
```
Comment (astralcai) [inline pennylane/typing.py:208]: So this would only work for `numpy.ndarray`-like objects? Essentially:
```pycon
>>> a = AbstractArray((-1, 2), int)
>>> isinstance([[1, 0], [0, 1]], a)
False
```

Comment (mudit2812) [reply pennylane/typing.py:208]: Yeah. But I didn't change that logic. Tbh I think we should just get rid of the instance check, but that is a discussion for another day. `is_compatible_with` is a much friendlier check

**Review by astralcai — APPROVED** [review]
LGTM

File: tests/math/test_init.py [hunk 9]
```diff
@@ -17,7 +17,7 @@
 import pytest
 
 import pennylane as qp
-from pennylane.templates.core import AbstractArray
+from pennylane.typing import AbstractArray
```
Comment (andrijapau) [inline tests/math/test_init.py:20]: lol

File: pennylane/typing.py (hunk 7 above)
Comment (mudit2812) [reply pennylane/typing.py:191]: I was doing that initially, but felt weird about implicitly looping over the shape twice instead of explicitly just once. That being said, the performance difference is probably negligible in real workflows. I'm sure we will see a bot PR in the future talking about performance loss here though 😆 

In any case, I'll apply your suggestion. The readability improvement is worth the negligible performance difference.

File: pennylane/typing.py [hunk 10]
```diff
@@ -137,79 +137,161 @@ class AbstractArray:
...
+
+    def is_compatible_with(self, val) -> bool:
+        """Check whether an input value is compatible with an ``AbstractArray``."""
```
Comment (andrijapau) [inline pennylane/typing.py]: Do you mind adding a quick docstring with examples here? I imagine this will be referenced a decent amount.

**Review by andrijapau — APPROVED** [review]
These improvements look good to me. Just a minor suggestion 🎉

Thanks! 🚀

File: pennylane/typing.py (hunk 10 above)
Comment (mudit2812) [reply pennylane/typing.py]: [f3e06aa](https://github.com/PennyLaneAI/pennylane/pull/9712/commits/f3e06aa6eec5e7c613b81cd85d1615045e667512)

**Review by andrijapau — APPROVED** [review]
